controller/machine/controller.go

```diff
@@ -300,3 +310,24 @@ func (m *Lifecycle) saveConfig(config *machineconfig.MachineConfig, machineDir s
 
 	return obj, nil
 }
+
+func (m *Lifecycle) isNodeInAppliedSpec(machine *v3.Machine) (bool, error) {
+	cluster, err := m.clusterLister.Get("", machine.Spec.ClusterName)
+	if err != nil {
+		return false, err
+	}
+	if cluster == nil {
+		return false, nil
+	}
+	if cluster.Status.AppliedSpec.RancherKubernetesEngineConfig == nil {
+		return false, nil
+	}
+
+	for _, node := range cluster.Status.AppliedSpec.RancherKubernetesEngineConfig.Nodes {
+		machineName := strings.Split(node.MachineName, ":")
+		if machineName[1] == machine.Name {
```

Comment:
can machinename ever be empty? lets guard against it, and if len(splitted) < 2, continue